cookbooks/CPO_induced_anisotropic_viscosity/plugin/cpo_induced_anisotropic_viscosity.cc

```diff
@@ -0,0 +1,587 @@
+              A[3][3] = 2.0/3.0*L;
+              A[4][4] = 2.0/3.0*M;
+              A[5][5] = 2.0/3.0*N;
+
+              // Invert using ScaLAPACK in dealii
+              FullMatrix<double> A_mat(6, 6);
+              for (unsigned int ai=0; ai<6; ++ai)
+                {
+                  for (unsigned int aj=0; aj<6; ++aj)
+                    {
+                      A_mat(ai,aj) = A[ai][aj];
+                    }
+                }
+
+              // A is the anisotropic tensor for the fluidity. We need its inverse, but it's not invertible due to singularity.
+              // Thus we use a pseudo inverse function imported from the scalapack package from deal.ii
+              const double ratio = 1e-8;
+              std::shared_ptr<Utilities::MPI::ProcessGrid> grid = std::make_shared<Utilities::MPI::ProcessGrid>(this->get_mpi_communicator(),6,6,4,4);
+              ScaLAPACKMatrix<double> A_scalapack(6,6,grid,4,4);
+              A_scalapack = A_mat;
+              A_scalapack.pseudoinverse(ratio);
+              FullMatrix<double> pinvA_mat(6,6);
+              A_scalapack.copy_to(pinvA_mat);
+
+              SymmetricTensor<2,6> invA;
+              for (unsigned int ai=0; ai<6; ++ai)
+                {
+                  for (unsigned int aj=0; aj<6; ++aj)
+                    {
+                      invA[ai][aj] = pinvA_mat(ai,aj);
+                    }
+                }
+
+              // Calculate the fluidity tensor in the CPO frame
+              Tensor<2,6> V = R_CPO_K * invA * transpose(R_CPO_K);//invA;//
+
+              // Convert rank 2 viscosity tensor to rank 4
+              FullMatrix<double> V_mat(6,6);
+              for (unsigned int vi=0; vi<6; ++vi)
+                {
+                  for (unsigned int vj=0; vj<6; ++vj)
+                    {
+                      V_mat[vi][vj] = V[vi][vj];
+                    }
+                }
+              SymmetricTensor<4,dim> V_r4;
+              dealii::Physics::Notation::Kelvin::to_tensor(V_mat, V_r4);
+
+              if (anisotropic_viscosity != nullptr)
+                {
+                  anisotropic_viscosity->stress_strain_directors[q] = V_r4;
+                }
+
+              double scalar_viscosity = composition[ind_vis];
+
+              // In first time step using input viscosity can lead to convergence issue if the strainrate varies significantly within the model domain.
+              // Thus for the first timestep we calculate an initial viscosity based on the strain rate. Why not later: seems to cause unstable solution(?)
+              if (this->get_timestep_number() == 1)
+                {
+                  const double edot_ii=std::max(std::sqrt(std::max(-second_invariant(deviator(strain_rate)), 0.)),
+                                                min_strain_rate);
+                  scalar_viscosity= 1/Gamma *  std::pow(edot_ii,((1. - n)/n));
+                }
+
+              double n_iterations = 1;
```

Comment:
Why not `unsigned int` and why start with 1?

```suggestion
              unsigned int n_iterations = 0;
```